### PR TITLE
feat: ai rule gen

### LINKED ISSUE #1987: [➕ Feature]: AI Creating Correlation Rules

### PR DESCRIPTION
**Pull request by mirrormystic**
close #1987

demo: https://www.loom.com/share/2b3eacfd7b844688bddaae6f2df7d8b6?sid=c07e0832-13e1-4249-98db-aa3c696c7b32

### CODE REVIEW

File: keep-ui/app/rules/AIGenRules.tsx [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+        )}
+      </div>
+    </TableHeaderCell>
+  );
+};
+
+export const AIGenRules: React.FC = () => {
+  const {triggerGenRules } = useGenRules();
+  const { data: session } = useSession();
+  const [sorting, setSorting] = React.useState<SortingState>([]);
```
Comment (ronshapiro) [inline keep-ui/app/rules/AIGenRules.tsx:79]: nit: just `useState()`

**Comment by mirrormystic** [comment]
https://www.loom.com/share/2b3eacfd7b844688bddaae6f2df7d8b6?sid=c07e0832-13e1-4249-98db-aa3c696c7b32   - feature preview

File: keep-ui/app/rules/AIGenRules.tsx [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+
+const SortableHeaderCell: React.FC<SortableHeaderCellProps> = ({
+  header,
+  children,
+}) => {
+  const { column } = header;
+
+  return (
+    <TableHeaderCell
+      className={`relative ${
```
Comment (ronshapiro) [inline keep-ui/app/rules/AIGenRules.tsx:37]: How about:
`className={`relative ${column.getIsPinned() ? "" : "hover:bg-slate-100"} group`}`

File: keep/api/routes/rules.py [hunk 3]
```diff
@@ -214,3 +228,206 @@ async def update_rule(
...
+
+        logger.info("Got {} rules back from the llm".format(len(result['results'])))
+
+        result['results'] = [rule for rule in result['results'] if check_cel_rule(rule['CELRule'])]
+
+        
+        pusher_client = get_pusher_client()
+        if pusher_client:      
+            try:
+                pusher_client.trigger("private-{}".format(tenant_id), "rules-aigen-created", result)
```
Comment (ronshapiro) [inline keep/api/routes/rules.py:391]: nit: `f"private-{tenant_id}"` is how I see things being done in the code.

Thinking out loud: does this trigger a websocket message to every user logged in with the same tenant?

Comment (mirrormystic) [reply keep/api/routes/rules.py:391]: thats a great point, i stole pattern from some other parts of the codebase, should we do something about this? @shahargl

File: keep-ui/app/rules/AIGenRules.tsx (hunk 2 above)
Comment (mirrormystic) [reply keep-ui/app/rules/AIGenRules.tsx]: fixed

**Comment by Kiryous** [comment]
Now if I don't have correlation rules yet I can't generate rules with AI. I think it's the perfect use case for the feature, so let's maybe always show the tabs with "existing" / "suggested" and add button "Suggest correlations with AI" which would get user to the "Suggested" tab. What do you think?
 
Screenshot 2024-10-13 at 15 06 56

File: keep-ui/app/rules/AIGenRules.tsx [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+    
+    mutateAIGeneratedRules(serverGenRules);
+  }, [serverGenRules]);
+
+  const handleGenerateMoreRules = () => {
+    setIsLoadingRules(true);
+    triggerGenRules();
+  };
+
+  const handleAddRule = async (rule: AIGeneratedRule) => {
```
Comment (Kiryous) [inline keep-ui/app/rules/AIGenRules.tsx:108]: Let's move `addRule()` function to `useRules()` hook to isolate data logic to the hook and this component to mostly UI.

File: keep-ui/app/rules/CorrelationSidebar/DeleteRule.tsx [hunk 5]
```diff
@@ -37,7 +37,7 @@ export const DeleteRuleCell = ({ ruleId }: DeleteRuleCellProps) => {
 
   return (
     <Button
-      className="invisible group-hover:visible"
+      className=""
```
Comment (Kiryous) [inline keep-ui/app/rules/CorrelationSidebar/DeleteRule.tsx:40]: leftover?

File: keep-ui/app/rules/AIGenRules.tsx [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+
+  const mutateAIGeneratedRules = (rules: AIGeneratedRule[]) => {
+    setGeneratedRules(rules);
+    setIsLoadingRules(false);
+  };
+
+  const { data:serverGenRules } = useRulePusherUpdates();
+
+  useEffect(() => {
+    console.log("serverGenRules", serverGenRules);
```
Comment (Kiryous) [inline keep-ui/app/rules/AIGenRules.tsx:94]: leftover

File: keep-ui/app/rules/AIGenRules.tsx [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+    getCoreRowModel: getCoreRowModel(),
+    getSortedRowModel: getSortedRowModel(),
+    onSortingChange: setSorting,
+    state: {
+      sorting,
+    },
+  });
+
+  if (isLoadingRules) {
+    return <Loading loadingText="Generating AI recommendations..." />;
```
Comment (Kiryous) [inline keep-ui/app/rules/AIGenRules.tsx:241]: I think it worth saying it would take a few minutes

File: keep-ui/app/rules/AIGenRules.tsx [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+      header: "Group By",
+      cell: (info) => info.getValue().join(", "),
+    }),
+    columnHelper.accessor("ChainOfThought", {
+      header: "Explanations",
+      cell: (info) => {
+        const rule = info.row.original;
+        return (
+          <div className="flex space-x-2">
+            <Icon
```
Comment (Kiryous) [inline keep-ui/app/rules/AIGenRules.tsx:175]: to fix tooltip, try to add
```suggestion
            <Icon
             key={rule.ChainOfThought}
```

my guess is that at first render content is not yet here but dimensions are already set.

**Review by Kiryous — CHANGES_REQUESTED** [review]
Awesome work, Haim! Left a suggestion to separate data managing and UI, and few minor comments.

File: keep-ui/app/rules/CorrelationSidebar/DeleteRule.tsx (hunk 5 above)
Comment (mirrormystic) [reply keep-ui/app/rules/CorrelationSidebar/DeleteRule.tsx]: fixed

File: keep/api/routes/rules.py [hunk 9]
```diff
@@ -214,3 +228,210 @@ async def update_rule(
...
+        IdentityManagerFactory.get_auth_verifier(["read:rules", "read:alerts"])
+    ),
+):
+    
+    task_id = str(uuid.uuid4())
+    background_tasks.add_task(ruleGen, task_id, authenticated_entity)
+    return {"task_id": task_id}
+
+
+def ruleGen(task_id, authenticated_entity):
```
Comment (Matvey-Kuk) [inline keep/api/routes/rules.py:349]: CamelCase, but the rest is snake_case ;)

**Review by Matvey-Kuk — COMMENTED** [review]
Sorry to ask 😅, but may I ask to craft some unit-tests?

File: keep/api/routes/rules.py [hunk 10]
```diff
@@ -12,6 +20,12 @@
 from keep.identitymanager.authenticatedentity import AuthenticatedEntity
 from keep.identitymanager.identitymanagerfactory import IdentityManagerFactory
 
+from keep.api.core.db import get_last_alerts
+from keep.api.core.dependencies import get_pusher_client
+
+# Add this import at the top of the file
```
[the remark this answers is not in the document]
Comment (mirrormystic) [reply keep/api/routes/rules.py:26]: done

File: keep/api/routes/rules.py [hunk 11]
```diff
@@ -214,3 +228,210 @@ async def update_rule(
...
+    try:
+        pusher_client = get_pusher_client()
+        if pusher_client:
+            pusher_client.trigger("private-{}".format(tenant_id), "rules-aigen-created", result)
+    except Exception as e:
+        logger.error(f"Error triggering Pusher event: {e}")
+
+
+
+def select_right_num_alerts(existing_rules, alerts, max_tokens):
```
[the remark this answers is not in the document]
Comment (mirrormystic) [reply keep/api/routes/rules.py:403]: good point, done

File: keep/api/routes/rules.py [hunk 12]
```diff
@@ -214,3 +228,210 @@ async def update_rule(
...
+
+        logger.info("Got {} rules back from the llm".format(len(result['results'])))
+
+        result['results'] = [rule for rule in result['results'] if check_cel_rule(rule['CELRule'])]
+    
+    except Exception as e:
+        logger.info(f"Error generating rules: {e}")
+        result = {'error': "Error generating rules"}
+
+
```
[the remark this answers is not in the document]
Comment (mirrormystic) [reply keep/api/routes/rules.py:393]: it just creates more clarity i feel

File: keep/api/routes/rules.py [hunk 13]
```diff
@@ -214,3 +228,210 @@ async def update_rule(
...
+            functions = RESULT_CUSTOM_FUNCTION,
+            function_call = 'auto'
+        )
+        result = json.loads(response.choices[0].message.function_call.arguments)
+
+        logger.info("Got {} rules back from the llm".format(len(result['results'])))
+
+        result['results'] = [rule for rule in result['results'] if check_cel_rule(rule['CELRule'])]
+    
+    except Exception as e:
```
[the remark this answers is not in the document]
Comment (mirrormystic) [reply keep/api/routes/rules.py:389]: Arq?, this wont create a retry, it just catchs all sorts of exceptions cuz there's a lof different things going on in the try, catch, i don't really want to deal with them in a case by case basis

File: keep/api/routes/rules.py [hunk 14]
```diff
@@ -214,3 +228,210 @@ async def update_rule(
...
+@router.get(
+    "/gen_rules",
+    description="Generate Rules Using An AI",
+    status_code=202,
+)
+def gen_rules(
+    background_tasks: BackgroundTasks,
+    request: Request,
+    authenticated_entity: AuthenticatedEntity = Depends(
+        IdentityManagerFactory.get_auth_verifier(["read:rules", "read:alerts"])
```
[the remark this answers is not in the document]
Comment (mirrormystic) [reply keep/api/routes/rules.py:340]: we're not reading incidents right now i think

File: keep/api/routes/rules.py [hunk 15]
```diff
@@ -214,3 +228,210 @@ async def update_rule(
...
+        if current_tokens + alert_tokens > available_tokens:
+            break
+        
+        selected_alerts.append(alert)
+        current_tokens += alert_tokens
+    
+    return selected_alerts
+
+
+def check_cel_rule(rule_str):
```
[the remark this answers is not in the document]
Comment (mirrormystic) [reply keep/api/routes/rules.py:428]: done

**Comment by Matvey-Kuk** [comment]
@mirrormystic I see your comments "done" but I con't see commits, is there a chance that you didn't push smth? 🤔

**Pull request closed without merging**